### PR TITLE
Internal function calls

### PR DESCRIPTION
**Pull request by MTrent1**
Interpret internal function calls.

### CODE REVIEW

File: src/interp/mod.rs [hunk 1]
```diff
@@ -212,6 +218,35 @@ impl LLVMIRInterpreter {
                         Name(n) => {
                             if n.as_str() == "printf" {
                                 self.printf(call);
+                            } else {
+                                match self.module.get_func_by_name(n) {
+                                    Some(func) => {
+                                        self.func_vars_stack.push(self.vars.clone());
```
Comment (ptersilie) [inline src/interp/mod.rs]: We can do without cloning `vars` here I think:
```
self.func_vars_stack.push(self.vars);
self.vars = HashMap::new();
```

File: src/interp/mod.rs [hunk 2]
```diff
@@ -212,6 +218,35 @@ impl LLVMIRInterpreter {
...
+                                        self.vars.clear();
+
+                                        let func = func.clone();
+                                        for (par, arg) in
+                                            func.parameters.iter().zip(call.arguments.iter())
+                                        {
+                                            self.vars.insert(par.name.clone(), arg.0.clone());
+                                        }
+
+                                        let value = self.it_bb(&func);
```
Comment (ptersilie) [inline src/interp/mod.rs]: We have to be careful with recursion again here.

File: lang_tests/inter_func_call.ll [hunk 3]
```diff
@@ -0,0 +1,71 @@
+; ModuleID = 'inter_func_call.c'
```
Comment (ptersilie) [inline lang_tests/inter_func_call.ll]: We should stop pushing and then later removing these files. They don't really add anything to the PR.

File: src/interp/mod.rs [hunk 4]
```diff
@@ -21,32 +21,36 @@ enum BinOps {
     Rem,
 }
 pub(crate) struct LLVMIRInterpreter {
+    module: Module,
+    func_vars_stack: Vec<HashMap<name::Name, Operand>>,
```
Comment (ptersilie) [inline src/interp/mod.rs]: We can just name this `callstack`.

File: src/interp/mod.rs [hunk 5]
```diff
@@ -57,11 +61,14 @@ impl LLVMIRInterpreter {
...
-        while bb_name != None {
-            //PERF: get_bb_by_name function is inefficient.
-            let bb = func.get_bb_by_name(bb_name.as_ref().unwrap()).unwrap();
+    fn it_bb(&mut self, func: &Function) -> Option<Operand> {
+        // for par in &func.parameters {
+        //     println!("cray cray")
+        // }
+
+        let mut bb_option = Some(&func.basic_blocks[0]);
+        while bb_option != None {
```
Comment (ptersilie) [inline src/interp/mod.rs]: `!bb_option.is_none()`

Comment (ptersilie) [reply src/interp/mod.rs]: Ah you are unwrapping immediately after. In that case we can do `while let Some(bb) = bb_option`.

File: src/interp/mod.rs (hunk 1 above)
Comment (MTrent1) [reply src/interp/mod.rs]: Doesn't seem to work as 'cannot move out of self.vars which is behind a mutable reference.

File: src/interp/mod.rs (hunk 2 above)
Comment (MTrent1) [reply src/interp/mod.rs]: But these are function calls and we want to return to the point where the function was called?

File: lang_tests/inter_func_call.ll (hunk 3 above)
Comment (MTrent1) [reply lang_tests/inter_func_call.ll]: Yes, sorry, I just forget to remove them.

File: src/interp/mod.rs (hunk 1 above)
Comment (ptersilie) [reply src/interp/mod.rs]: I see. I still don't feel good about cloning vars here when it's not strictly necessary. There are other ways to work around this, e.g. by not having `self.vars` at all, and make inserting a new variable always insert it into the map on top of the call stack (see `Vec::last_mut`). I'd probably refactor var inserting into its own function to do that.

Comment (ltratt) [reply src/interp/mod.rs]: Would e.g. `Rc` help here? [Or... Jake's `Gc`?]

File: src/interp/mod.rs (hunk 2 above)
Comment (ptersilie) [reply src/interp/mod.rs]: We do. That doesn't mean it needs to be recursive though. We can just remember where we are before doing the call (by storing func, block, instruction index), and then continue interpreting from there after we've seen a return.

File: lang_tests/inter_func_call.ll (hunk 3 above)
Comment (ptersilie) [reply lang_tests/inter_func_call.ll]: You might want to only add files you want to push in the first place, then you don't need to remove them. Just do `git add filename` for each file you want to add or use a graphical tool like `gitk`.

File: src/interp/mod.rs (hunk 1 above)
Comment (ptersilie) [reply src/interp/mod.rs]: Yes, that might also work.

Comment (MTrent1) [reply src/interp/mod.rs]: Rc?

Comment (ptersilie) [reply src/interp/mod.rs]: https://doc.rust-lang.org/rust-by-example/std/rc.html

Comment (ltratt) [reply src/interp/mod.rs]: https://doc.rust-lang.org/std/rc/struct.Rc.html -- it allows you to share things, although mutating them is difficult. It's probably not worth fixing in this PR, but worth looking at for the next. You might also want to ask @jacob-hughes about his `Gc` type, which is more suitable for interpreters (but does mean you can't use normal rustc at the moment).

Comment (ptersilie) [reply src/interp/mod.rs]: > It's probably not worth fixing in this PR

Should we fix the unnecessary clone in the meantime though?

Comment (ltratt) [reply src/interp/mod.rs]: If the clone is unnecessary, we should definitely kill it! [`cargo clippy` might also be worth running at some point soon -- it often spots unnecessary `clone`s.]

**Comment by MTrent1** [comment]
Ready for iterative solution review.

**Comment by MTrent1** [comment]
Ready for re-review.

**Comment by ptersilie** [comment]
Looking good. Please squash.

**Comment by MTrent1** [comment]
Ready for re-review.

**Comment by ptersilie** [comment]
Please squash.

**Comment by MTrent1** [comment]
Squashed.

**Pull request merged**